core/src/main/scala/org/elasticmq/actor/queue/operations/MoveMessagesAsyncOps.scala

```diff
@@ -0,0 +1,159 @@
+
+sealed trait MessageMoveTaskState
+case object NotMovingMessages extends MessageMoveTaskState
+case class MovingMessagesInProgress(
+    numberOfMessagesMoved: Long,
+    numberOfMessagesToMove: Long,
+    destinationArn: Option[String],
+    maxNumberOfMessagesPerSecond: Option[Int],
+    sourceArn: String,
+    startedTimestamp: Long,
+    taskHandle: MessageMoveTaskHandle
+) extends MessageMoveTaskState
+
+case class MessageMoveTaskData(
+    numberOfMessagesMoved: Long,
+    numberOfMessagesToMove: Long,
+    destinationArn: Option[String],
+    maxNumberOfMessagesPerSecond: Option[Int],
+    sourceArn: String,
+    startedTimestamp: Long,
+    status: String, // RUNNING, COMPLETED, CANCELLING, CANCELLED, and FAILED
+    taskHandle: MessageMoveTaskHandle
+)
+
+trait MoveMessagesAsyncOps extends Logging {
+  this: QueueActorStorage =>
+
+  private val prevMessageMoveTasks = collection.mutable.Buffer[MessageMoveTaskData]()
+  private var messageMoveTaskState: MessageMoveTaskState = NotMovingMessages
+
+  def startMovingMessages(
+      destinationQueue: ActorRef,
+      destinationArn: Option[String],
+      sourceArn: String,
+      maxNumberOfMessagesPerSecond: Option[Int],
+      queueManager: ActorRef
+  ): Either[ElasticMQError, MessageMoveTaskHandle] = {
+    messageMoveTaskState match {
+      case NotMovingMessages =>
+        val taskHandle = UUID.randomUUID().toString
+        logger.debug("Starting message move task to queue {} (task handle: {})", destinationQueue, taskHandle)
+        messageMoveTaskState = MovingMessagesInProgress(
+          0,
+          messageQueue.size,
+          destinationArn,
+          maxNumberOfMessagesPerSecond,
+          sourceArn,
+          startedTimestamp = System.currentTimeMillis(),
+          taskHandle
+        )
+        context.self ! MoveFirstMessage(destinationQueue, queueManager)
+        Right(taskHandle)
+      case _: MovingMessagesInProgress =>
+        Left(new MessageMoveTaskAlreadyRunning(queueData.name))
+    }
+  }
+
+  def moveFirstMessage(
+      destinationQueue: ActorRef,
+      queueManager: ActorRef
+  ): ResultWithEvents[Unit] = {
+    messageMoveTaskState match {
+      case NotMovingMessages =>
+        logger.debug("Not moving messages")
+        ResultWithEvents.empty
+      case mmInProgress @ MovingMessagesInProgress(
+            numberOfMessagesMoved,
+            _,
+            destinationArn,
+            maxNumberOfMessagesPerSecond,
+            sourceArn,
+            startedTimestamp,
+            taskHandle
+          ) =>
+        logger.debug("Trying to move a single message to {} ({} messages left)", destinationQueue, messageQueue.size)
+        messageQueue.pop match {
+          case Some(internalMessage) =>
+            messageMoveTaskState = mmInProgress.copy(numberOfMessagesMoved = numberOfMessagesMoved + 1)
+            destinationQueue ! SendMessage(internalMessage.toNewMessageData)
+            maxNumberOfMessagesPerSecond match {
```

Comment:
**question:** Do we need to implement the throttling? I think the `maxNumberOfMessagesPerSecond` may be relevant for real backends to avoid performance issues, but is it a case for ElasticMQ? Maybe users who use it as a stub would prefer to have this done immediately?
Edit: Ok, there may be a use case where users want to test their cancellation logic, so they would use this parameter in tests to emulate a slow cancellable process, like you did, right?

Reply:
Exactly, I'm using this feature in integration tests as well. Beside that, it's quite easy to emulate it with Akka.